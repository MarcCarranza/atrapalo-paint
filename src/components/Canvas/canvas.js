import React, { Component } from "react";
import { Stage, Layer, Line } from "react-konva";
import classes from "./canvas.module.css";

class Canvas extends Component {
  constructor(props) {
    super(props);
    this.state = {
      lines: [],        // Array donde guardar las lineas
      savedLines: [],   // Array donde guardar las lineas cuando se llama a deshacer
      linesConfig: [],  // Array donde guardar la configuracion de las lineas
      savedLinesConfig: [],               // Array donde guardar la configuracion de las lineas al deshacer
      width: window.innerWidth * 0.7,     // Variable para calcular el ancho del canvas
      height: window.innerHeight * 0.8    // Variable para calcular la altura del canvas
    };
  }

  // Añade un event listener que llama a canvasResize()
  componentDidMount() {
    window.addEventListener("resize", this.canvasResize());
  }

  // Comprueba si ha cambiado el estado de undo/redo y lo actualiza acorde a la acción (esto debe irse)
  componentDidUpdate(prevProps) {
    if (prevProps.undo < this.props.undo) {
      this.handleUndo();
    } else if (prevProps.undo > this.props.undo) {
      this.handleRedo();
    }
  }

  // Función para que el canvas sea responsive
  canvasResize = () => {
    this.setState({
      width: window.innerWidth * 0.7,
      height: window.innerHeight * 0.8
    });
  };

  // Función que activa o desactiva los botones en función de su disponibilidad
  enableUndoRedo = () => {
    let undo, redo;
    if (this.state.lines.length === 0) {
      undo = false;
    } else if (this.state.lines.length > 0) {
      undo = true;
    }
    if (this.state.savedLines.length === 0) {
      redo = false;
    } else if (this.state.savedLines.length > 0) {
      redo = true;
    }

    this.props.click(undo, redo);
  };

  // Función para organizar las acciones de deshacer
  handleUndo = () => {
    this.saveOldLines();
    this.saveOldLinesConfig();
  };

  // Función que guarda las lineas en otro array cuando llama a deshacer
  saveOldLines = () => {
    let newLines = this.state.lines;
    let lineToSave = newLines[newLines.length - 1];
    let oldLines = this.state.savedLines;
    oldLines.push(lineToSave);
    newLines.splice(newLines.length - 1, 1);
    this.setState({
      lines: newLines,
      savedLines: oldLines
    });
  };

  // Función que guarda la configuración de las lineas cuando se llama a deshacer
  saveOldLinesConfig = () => {
    let newLinesConfig = this.state.linesConfig;
    let lineConfigToSave = newLinesConfig[newLinesConfig.length -1];
    let oldLinesConfig = this.state.savedLinesConfig;
    oldLinesConfig.push(lineConfigToSave);
    newLinesConfig.splice(newLinesConfig.length - 1, 1);
    this.setState({
      linesConfig: newLinesConfig,
      savedLinesConfig: oldLinesConfig
    })
  }

  // Función para organizar las acciones de rehacer (por aquí he perdido la esperanza)
  handleRedo = () => {
    this.restoreOldLines();
    this.restoreOldLinesConfig();
  };

  // Función que restora las lineas cuando se llama a rehacer
  restoreOldLines = () => {
    if (this.props.undo >= 0) {
      let newLines = this.state.lines;
      let redoLine = this.state.savedLines[this.state.savedLines.length - 1];
      let oldLines = this.state.savedLines;
      newLines.push(redoLine);
      oldLines.splice(oldLines.length - 1, 1);
      this.setState({
        lines: newLines,
        savedLines: oldLines
      });
    }
  }

  // Función que restora la configuración de las lineas cuando se llama a rehacer
  restoreOldLinesConfig = () => {
    if (this.props.undo >= 0) {
      let newLinesConfig = this.state.linesConfig;
      let redoLineConfig = this.state.savedLinesConfig[this.state.savedLinesConfig.length - 1];
      let oldLinesConfig = this.state.savedLinesConfig;
      newLinesConfig.push(redoLineConfig);
      oldLinesConfig.splice(oldLinesConfig.length - 1, 1);
      this.setState({
        linesConfig: newLinesConfig,
        savedLinesConfig: oldLinesConfig
      });
    }
  }

  // Función que se encarga de detectar si el canvas está siendo presionado y hacer un spread para mostrar las lineas
  handleMouseDown = () => {
    this.saveLinesConfig();
    this._drawing = true;
    this.setState({
      lines: [...this.state.lines, []]
    });
  };

  // Función que se encarga de detectar el movimiento del ratón y guarda las lineas
  handleMouseMove = e => {
    this.enableUndoRedo();
    if (!this._drawing) {
      return;
    }
    const stage = this.stageRef.getStage();
    const point = stage.getPointerPosition();
    const { lines } = this.state;

    let lastLine = lines[lines.length - 1];
    lastLine = lastLine.concat([point.x, point.y]);

    lines.splice(lines.length - 1, 1, lastLine);
    this.setState({
      lines: lines.concat()
    });
  };

  // Función que detecta cuando se levanta el ratón
  handleMouseUp = () => {
    this._drawing = false;
  };

  // Guarda los datos (color, width) de las lineas en un array aparte
  saveLinesConfig = () => {
    let lineArray = this.state.linesConfig;
    let lineColor = this.props.strokeColor;
    let lineWidth = this.props.strokeWidth;
    let line = {
      lineColor,
      lineWidth
    };
    lineArray.push(line);
    this.setState({
      lineConfig: lineArray
    });
  };

  render() {
    return (
      <div className={classes.Canvas}>
        <Stage
          width={this.state.width}
          height={this.state.height}
          onContentMousedown={this.handleMouseDown}
          onContentMousemove={this.handleMouseMove}
          onContentMouseup={this.handleMouseUp}
          ref={node => {
            this.stageRef = node;
          }}
        >
          <Layer>
            {this.state.lines.map((line, i) => (
              <Line
                key={i}
                points={line}
                stroke={this.state.linesConfig[i].lineColor}
                strokeWidth={this.state.linesConfig[i].lineWidth}
              />
            ))}
          </Layer>
        </Stage>
      </div>
    );
  }
}

export default Canvas;
