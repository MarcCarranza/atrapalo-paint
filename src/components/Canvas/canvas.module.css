.Canvas{
    width: 70vw;
    height: 80vh;
    border: 1px solid #777;
    margin: 2rem;
}