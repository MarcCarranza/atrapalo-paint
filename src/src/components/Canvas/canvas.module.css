.CanvasWrapper{
    grid-area: canvas;
    display: flex;
    justify-content: center;
}

.Canvas{
    border: 1px solid var(--color-border);
    width: 80vw;
    height: 85vh;
    background-color: white;
    overflow: auto;
}

