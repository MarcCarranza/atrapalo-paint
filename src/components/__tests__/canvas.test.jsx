import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from "enzyme";
import Canvas from "../Canvas/canvas";

describe("Canvas", () => {
  Enzyme.configure({ adapter: new Adapter()})
  const clickTest = jest.fn();
  it("renders correctly", () => {
    const wrapper = Enzyme.shallow(
      <Canvas 
        click={clickTest}
        />
    );
    expect(wrapper).toMatchSnapshot();
    expect(clickTest).toHaveBeenCalled();
    // On the first run of this test, Jest will generate a snapshot file automatically.
  });
});
