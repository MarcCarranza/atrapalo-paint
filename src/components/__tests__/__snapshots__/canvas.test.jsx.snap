// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Canvas
    click={
      [MockFunction] {
        "calls": Array [
          Array [
            false,
            false,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ForwardRef
        className="Canvas"
        height={648.96}
        onContentMousedown={[Function]}
        onContentMousemove={[Function]}
        onContentMouseup={[Function]}
        width={814.08}
      >
        <Layer />
      </ForwardRef>,
      "className": "CanvasWrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Layer />,
        "className": "Canvas",
        "height": 648.96,
        "onContentMousedown": [Function],
        "onContentMousemove": [Function],
        "onContentMouseup": [Function],
        "width": 814.08,
      },
      "ref": [Function],
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
        },
        "ref": null,
        "rendered": Array [],
        "type": "Layer",
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ForwardRef
          className="Canvas"
          height={648.96}
          onContentMousedown={[Function]}
          onContentMousemove={[Function]}
          onContentMouseup={[Function]}
          width={814.08}
        >
          <Layer />
        </ForwardRef>,
        "className": "CanvasWrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Layer />,
          "className": "Canvas",
          "height": 648.96,
          "onContentMousedown": [Function],
          "onContentMousemove": [Function],
          "onContentMouseup": [Function],
          "width": 814.08,
        },
        "ref": [Function],
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": "Layer",
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
